src/linear_algebra/determinant.lean

```diff
@@ -103,4 +106,101 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
 { map_one := det_one,
   map_mul := det_mul }
 
+/-- Transposing a matrix preserves the determinant. -/
+@[simp] lemma det_transpose (M : matrix n n R) : M.transpose.det = M.det :=
+begin
+  apply sum_bij (λ σ _, σ⁻¹),
+  { intros σ _, apply mem_univ },
+  { intros σ _,
+    rw [sign_inv],
+    congr' 1,
+    apply prod_bij (λ i _, σ i),
+    { intros i _, apply mem_univ },
+    { intros i _, simp },
+    { intros i j _ _ h, simp at h, assumption },
+    { intros i _, use σ⁻¹ i, finish }
+  },
+  { intros σ σ' _ _ h, simp at h, assumption },
+  { intros σ _, use σ⁻¹, finish }
+end
+
+/-- The determinant of a permutation matrix equals its sign. -/
+@[simp] lemma det_permutation (σ : perm n) :
+  matrix.det (σ.to_pequiv.to_matrix : matrix n n R) = σ.sign := begin
+  suffices : matrix.det (σ.to_pequiv.to_matrix) = ↑σ.sign * det (1 : matrix n n R), { simp [this] },
+  unfold det,
+  rw mul_sum,
+  apply sum_bij (λ τ _, σ * τ),
+  { intros τ _, apply mem_univ },
+  { intros τ _,
+    conv_lhs { rw [←one_mul (sign τ), ←int.units_pow_two (sign σ)] },
+    conv_rhs { rw [←mul_assoc, coe_coe, sign_mul, units.coe_mul, int.cast_mul, ←mul_assoc] },
+    congr,
+    { norm_num },
+    { ext i, apply pequiv.equiv_to_pequiv_to_matrix }
+  },
+  { intros τ τ' _ _, exact (mul_left_inj σ).mp },
+  { intros τ _, use σ⁻¹ * τ, use (mem_univ _), exact (mul_inv_cancel_left _ _).symm }
+end
+
+/-- Permuting the columns changes the sign of the determinant. -/
+lemma det_permute (σ : perm n) (M : matrix n n R) : matrix.det (λ i, M (σ i)) = σ.sign * M.det :=
+by rw [←det_permutation, ←det_mul, pequiv.to_pequiv_mul_matrix]
+
+section det_zero
+/-! ### `det_zero` section
+
+  Prove that a matrix with a repeated column has determinant equal to zero.
+-/
+
+/--
+  `mod_swap i j` contains permutations up to swapping `i` and `j`.
+
+  We use this to partition permutations in the expression for the determinant,
+  such that each partitions sums up to `0`.
+-/
+def mod_swap {n : Type u} [decidable_eq n] (i j : n) : setoid (perm n) :=
+⟨ λ σ τ, σ = τ ∨ σ = swap i j * τ
+, λ σ, or.inl (refl σ)
+, λ σ τ h, or.cases_on h
+(λ h, or.inl h.symm)
+(λ h, or.inr (by rw [h, swap_mul_self_mul]))
+, λ σ τ υ hστ hτυ, by cases hστ; cases hτυ; try {rw [hστ, hτυ, swap_mul_self_mul]}; finish
+⟩
+
+instance (i j : n) : decidable_rel (mod_swap i j).r := λ σ τ, or.decidable
+
+variables {M : matrix n n R} {i j : n} (i_ne_j : i ≠ j) (hij : M i = M j)
```

Comment:
could you include the Prop variables `i_ne_j` and `hij` in the statements of the lemmas instead of the variable list, to help the readers understand the statements?